File: python/listas/dobleLista.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None
        self.ant = None


head = None
tail = None


def beg_insert():
    global head, tail

    item = int(input("\nIngrese valor: "))
    new_node = Node(item)

    new_node.next = head
    new_node.ant = None

    if head is not None:
        head.ant = new_node
    else:
        tail = new_node

    head = new_node
    print("Nodo insertado al inicio")


def last_insert():
    global head, tail

    item = int(input("\nIngrese valor: "))
    new_node = Node(item)

    if head is None:
        head = new_node
        tail = new_node
    else:
        tail.next = new_node
        new_node.ant = tail
        tail = new_node

    print("Nodo insertado al final")


def select_insert():
    global head, tail

    item = int(input("\nIngrese el valor: "))
    loc = int(input("Introduce la ubicación después de la cual deseas ingresar: "))

    if loc <= 0:
        print("Ubicación no válida")
        return

    if head is None:
        print("No se puede insertar")
        return

    temp = head
    for i in range(1, loc):
        if temp.next is None:
            print("No se puede insertar")
            return
        temp = temp.next

    new_node = Node(item)

    new_node.next = temp.next
    new_node.ant = temp

    if temp.next is not None:
        temp.next.ant = new_node
    else:
        tail = new_node

    temp.next = new_node
    print("Nodo insertado")


def begin_delete():
    global head, tail

    if head is None:
        print("Lista vacía")
    elif head.next is None:
        head = None
        tail = None
        print("Único nodo eliminado")
    else:
        head = head.next
        head.ant = None
        print("Primer nodo eliminado")


def last_delete():
    global head, tail

    if head is None:
        print("Lista vacía")
    elif tail.ant is None:
        head = None
        tail = None
        print("Único nodo eliminado")
    else:
        tail = tail.ant
        tail.next = None
        print("Último nodo eliminado")


def select_delete():
    global head, tail

    loc = int(input("\nIntroduce la ubicación después de la cual deseas eliminar: "))

    if head is None:
        print("Lista vacía, no se puede eliminar")
        return

    if loc <= 0:
        print("Ubicación inválida")
        return

    ptr = head
    prev = None

    for i in range(loc):
        prev = ptr
        ptr = ptr.next
        if ptr is None:
            print("Ubicación no encontrada")
            return

    prev.next = ptr.next

    if ptr.next is not None:
        ptr.next.ant = prev
    else:
        tail = prev

    print(f"Nodo eliminado en la posición {loc+1}")


def search():
    if head is None:
        print("Lista vacía")
        return

    item = int(input("\nIntroduce el elemento que deseas buscar: "))

    ptr = head
    pos = 1
    found = False

    while ptr is not None:
        if ptr.data == item:
            print(f"Elemento encontrado en la ubicación {pos}")
            found = True
        ptr = ptr.next
        pos += 1

    if not found:
        print("Elemento no encontrado")


def display():
    if head is None:
        print("Nada que imprimir")
        return

    print("\nImprimiendo valores...")
    ptr = head
    while ptr is not None:
        print(ptr.data)
        ptr = ptr.next


# -------- MENÚ PRINCIPAL --------
def menu():
    choice = 0

    while choice != 9:
        print("\n********** Menú principal **********")
        print("1. Insertar al principio")
        print("2. Insertar al final")
        print("3. Insertar después de una ubicación")
        print("4. Eliminar del principio")
        print("5. Eliminar desde el último")
        print("6. Eliminar después de una ubicación")
        print("7. Buscar un elemento")
        print("8. Mostrar")
        print("9. Salir")

        choice = int(input("\nIngrese su opción: "))

        if choice == 1:
            beg_insert()
        elif choice == 2:
            last_insert()
        elif choice == 3:
            select_insert()
        elif choice == 4:
            begin_delete()
        elif choice == 5:
            last_delete()
        elif choice == 6:
            select_delete()
        elif choice == 7:
            search()
        elif choice == 8:
            display()
        elif choice == 9:
            print("Saliendo...")
        else:
            print("Opción inválida")


# Ejecutar el programa
menu()

File: javascript/listas/dobleLista.js
const readline = require("readline");

class Node {
  constructor(data) {
    this.data = data;
    this.next = null;
    this.ant = null; // anterior
  }
}

let head = null;
let tail = null;

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout
});

function menu() {
  console.log("\n********** Menú principal **********");
  console.log("1. Insertar al principio");
  console.log("2. Insertar al final");
  console.log("3. Insertar después de ubicación");
  console.log("4. Eliminar del principio");
  console.log("5. Eliminar desde el último");
  console.log("6. Eliminar nodo después de la ubicación");
  console.log("7. Buscar un elemento");
  console.log("8. Mostrar");
  console.log("9. Salir");

  rl.question("\nIngrese su opción: ", (choice) => {
    switch (parseInt(choice)) {
      case 1: return begInsert();
      case 2: return lastInsert();
      case 3: return selectInsert();
      case 4: return beginDelete();
      case 5: return lastDelete();
      case 6: return selectDelete();
      case 7: return search();
      case 8: return display();
      case 9:
        rl.close();
        return;
      default:
        console.log("Opción inválida");
        return menu();
    }
  });
}

// INSERTAR AL INICIO
function begInsert() {
  rl.question("Ingrese valor: ", (value) => {
    const newNode = new Node(parseInt(value));

    newNode.next = head;
    newNode.ant = null;

    if (head !== null) {
      head.ant = newNode;
    } else {
      tail = newNode;
    }

    head = newNode;
    console.log("Nodo insertado al inicio");
    menu();
  });
}

// INSERTAR AL FINAL
function lastInsert() {
  rl.question("Ingrese valor: ", (value) => {
    const newNode = new Node(parseInt(value));

    if (head === null) {
      head = newNode;
      tail = newNode;
    } else {
      tail.next = newNode;
      newNode.ant = tail;
      tail = newNode;
    }

    console.log("Nodo insertado al final");
    menu();
  });
}

// INSERTAR DESPUÉS DE POSICIÓN
function selectInsert() {
  rl.question("Ingrese valor: ", (value) => {
    const item = parseInt(value);

    rl.question("Ingrese ubicación (posición): ", (loc) => {
      const position = parseInt(loc);

      if (position <= 0) {
        console.log("Ubicación no válida");
        return menu();
      }

      let temp = head;
      for (let i = 1; i < position; i++) {
        if (!temp) {
          console.log("No se puede insertar");
          return menu();
        }
        temp = temp.next;
      }

      if (!temp) {
        console.log("No se puede insertar");
        return menu();
      }

      const newNode = new Node(item);

      newNode.next = temp.next;
      newNode.ant = temp;

      if (temp.next !== null) {
        temp.next.ant = newNode;
      } else {
        tail = newNode;
      }

      temp.next = newNode;

      console.log("Nodo insertado");
      menu();
    });
  });
}

// ELIMINAR PRIMERO
function beginDelete() {
  if (head === null) {
    console.log("Lista vacía");
  } 
  else if (head.next === null) {
    head = null;
    tail = null;
    console.log("Único nodo eliminado");
  } 
  else {
    head = head.next;
    head.ant = null;
    console.log("Primer nodo eliminado");
  }
  menu();
}

// ELIMINAR ÚLTIMO
function lastDelete() {
  if (head === null) {
    console.log("Lista vacía");
  } 
  else if (tail.ant === null) {
    head = null;
    tail = null;
    console.log("Único nodo eliminado");
  } 
  else {
    tail = tail.ant;
    tail.next = null;
    console.log("Último nodo eliminado");
  }

  menu();
}

// ELIMINAR DESPUÉS DE POSICIÓN
function selectDelete() {
  rl.question("Ingrese posición: ", (value) => {
    const loc = parseInt(value);

    if (head === null) {
      console.log("Lista vacía");
      return menu();
    }

    if (loc <= 0) {
      console.log("La ubicación debe ser mayor que 0");
      return menu();
    }

    let ptr = head;
    let prev = null;

    for (let i = 0; i < loc; i++) {
      prev = ptr;
      ptr = ptr.next;

      if (ptr == null) {
        console.log("Ubicación no encontrada");
        return menu();
      }
    }

    prev.next = ptr.next;

    if (ptr.next !== null) {
      ptr.next.ant = prev;
    } else {
      tail = prev;
    }

    console.log("Nodo eliminado en la posición", loc + 1);
    menu();
  });
}

// BUSCAR
function search() {
  if (head === null) {
    console.log("Lista vacía");
    return menu();
  }

  rl.question("Ingrese valor a buscar: ", (value) => {
    const item = parseInt(value);

    let ptr = head;
    let pos = 1;
    let found = false;

    while (ptr !== null) {
      if (ptr.data === item) {
        console.log(`Elemento encontrado en la posición ${pos}`);
        found = true;
      }
      ptr = ptr.next;
      pos++;
    }

    if (!found) {
      console.log("Elemento no encontrado");
    }

    menu();
  });
}

// MOSTRAR
function display() {
  let ptr = head;

  if (ptr === null) {
    console.log("Nada que imprimir");
  } else {
    console.log("\nMostrando lista:");
    while (ptr !== null) {
      console.log(ptr.data);
      ptr = ptr.next;
    }
  }
  menu();
}

// Ejecutar
menu();
